crypto/evp/rand.c

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include "internal/cryptlib.h"
+#include <openssl/engine.h>
+#include <openssl/evp.h>
+#include <openssl/x509v3.h>
+#include <openssl/rand.h>
+#include <openssl/core.h>
+#include <openssl/core_names.h>
+#include "crypto/asn1.h"
+#include "crypto/evp.h"
+#include "internal/numbers.h"
+#include "internal/provider.h"
+#include "evp_local.h"
+
+EVP_RAND_CTX *EVP_RAND_CTX_new(EVP_RAND *rand)
+{
+    EVP_RAND_CTX *ctx = NULL;
+
+    if (rand == NULL)
+        return NULL;
+
+    ctx = OPENSSL_zalloc(sizeof(EVP_RAND_CTX));
+    if (ctx == NULL
+        || (ctx->data = rand->newctx(ossl_provider_ctx(rand->prov))) == NULL
+        || !EVP_RAND_up_ref(rand)) {
+        EVPerr(EVP_F_EVP_RAND_CTX_NEW, ERR_R_MALLOC_FAILURE);
+        if (ctx != NULL)
+            rand->freectx(ctx->data);
+        OPENSSL_free(ctx);
+        ctx = NULL;
+    } else {
+        ctx->meth = rand;
+    }
+    return ctx;
+}
+
+void EVP_RAND_CTX_free(EVP_RAND_CTX *ctx)
+{
+    if (ctx != NULL) {
+        ctx->meth->freectx(ctx->data);
+        ctx->data = NULL;
+        EVP_RAND_free(ctx->meth);
+        OPENSSL_free(ctx);
+    }
+}
+
+const char *EVP_RAND_name(const EVP_RAND *rand)
+{
+    return evp_first_name(rand->prov, rand->name_id);
+}
+
+const OSSL_PROVIDER *EVP_RAND_provider(const EVP_RAND *rand)
+{
+    return rand->prov;
+}
+
+const EVP_RAND *EVP_RAND_CTX_rand(EVP_RAND_CTX *ctx)
+{
+    return ctx->meth;
+}
+
+int EVP_RAND_CTX_reseed(EVP_RAND_CTX *ctx)
+{
+    if (ctx != NULL && ctx->meth->reseed != NULL)
+        return ctx->meth->reseed(ctx->data);
+    return 1;
+}
+
+unsigned int EVP_RAND_CTX_strength(EVP_RAND_CTX *ctx)
+{
+    OSSL_PARAM params[2] = { OSSL_PARAM_END, OSSL_PARAM_END };
+    unsigned int s;
+
+    if (ctx == NULL)
+        return 0;
+
+    *params = OSSL_PARAM_construct_uint(OSSL_RAND_PARAM_STRENGTH, &s);
+    if (ctx->meth->get_ctx_params != NULL
+        && ctx->meth->get_ctx_params(ctx->data, params))
+            return s;
+    if (ctx->meth->get_params != NULL
+        && ctx->meth->get_params(params))
+            return s;
+    return 0;
+}
+
+int EVP_RAND_CTX_get_bytes(EVP_RAND_CTX *ctx, void *out, size_t outlen)
+{
+    if (outlen == 0)
+        return 1;
+    if (ctx == NULL || out == NULL)
+        return 0;
+    return ctx->meth->get_bytes(ctx->data, out, outlen);
+}
+
+/*
+ * The {get,set}_params functions return 1 if there is no corresponding
```

Comment:
maybe they should return 0, (or -1) for not supported?

Reply:
This is the same as is in `mac_lib.c` and `kdf_lib.c`.  It might not be ideal but it is consistent.

We need to canonically define what happens when some of the params are not known.